platforms/web/lib/dom.ts

```diff
@@ -389,3 +431,74 @@ export function countCodeunit(
         return -1;
     }
 }
+
+/**
+ * Given a text node, determine if we need to add an additional offset to it. A
+ * text node that has any ancestor that is a li, pre, blockquote or p tag will
+ * require an additional offset to match up with the rust model. We also need to
+ * handle the case where we have consecutive formatting nodes, as only the last
+ * of the formatting nodes needs the extra offset (if applicable).
+ *
+ * This is called in two places and only with TextNodes
+ *
+ * Returns a boolean, true if the node needs an extra offset
+ */
+
+function nodeNeedsExtraOffset(node: Node | null) {
+    if (node === null) return false;
+
+    let checkNode: Node | ParentNode | null = node;
+    const hasFormattingParent = isFormattingNode(checkNode.parentNode);
+
+    // If the parent is _not_ a formatting node, then we have a case where the
+    // text node we are looking at is in a container, so we simply need to check
+    // the ancestors to see if one of those containers requires an extra offset.
+
+    // Otherwise we are dealing with the case where we are inside at least one
+    // formatting node (but we could be deeper than that) and we also need to
+    // make sure that we don't add the offset more than once when we have
+    // multiple adjacent inline formatting nodes.
+    while (checkNode) {
+        // if we have a formatting ancestor and the next sibling is not a
+        // container node, stop looking (and return false)
+        const nextSibling = checkNode.nextSibling;
+        if (
+            hasFormattingParent &&
+            nextSibling &&
+            !isNodeRequiringExtraOffset(nextSibling)
+        ) {
+            break;
+        }
+
+        if (isNodeRequiringExtraOffset(checkNode)) {
+            return true;
+        } else {
+            checkNode = checkNode.parentNode;
+        }
+    }
+    return false;
+}
+
+function isFormattingNode(node: Node | ParentNode | null) {
+    if (node === null) return false;
+    const formattingNodeNames = ['EM', 'U', 'STRONG', 'DEL'];
```

Comment:
I wonder if link nodes or inline code should also be included here, as they behave almost like formatting nodes.

Reply:
Interesting, I'll take a look at that.

If I expand this list of node names to include `a` and `code`, am I just at the point where I'm saying `!isCountainerNode(...)`. I'll have a look at the node types in the rust model and see if I can figure that out. If that's the case that could slim this PR down a bit.

Reply:
This seems to be working fine without as it stands, although I can't test links in the example app at the moment. I'll bear this in mind when I'm integrating into element web when I play with links

Reply:
... I take that back. Figured out how to break it. Have added the tags!